### DISCUSSION TITLE
Dashy update seems broken

### DISCUSSION THREAD

**Discussion by yuriw** [tteck/Proxmox#2631] [Help Needed]
After running `update` for Dashy got "No Dashy Installation Found!"

Installed a fresh Dash LXC, Debian 12 default and tried `update` after it,  got the same error.

**Comment by tteck** [comment]
https://github.com/tteck/Proxmox/issues/2614

**Reply by tteck** [reply]
https://github.com/tteck/Proxmox/commit/321660487de3d1df302c47e0ba21a999db28020b https://github.com/tteck/Proxmox/commit/183340bc853089e0a3800a4da527b6b9614bb175

**Reply by yuriw** [reply]
@tteck thx for quick response 
I wonder when the script runs from the container, will it get the latest code?
Or you implied that I have to ‘git pull’ the latest?

**Reply by tteck** [reply]
If you had read https://github.com/tteck/Proxmox/issues/2614#issuecomment-1982036903, you would have learned that the "latest" (2.1.2) breaks the installation.

https://github.com/tteck/Proxmox/blob/321660487de3d1df302c47e0ba21a999db28020b/ct/dashy.sh#L58

**Comment by yuriw** [comment]
Update is still broken 😞

**Reply by tteck** [reply]
https://github.com/tteck/Proxmox/commit/58b75e0dd8096ea4225e1c1ba8de2bd32a7d86f6

**Comment by yuriw** [comment]
I don’t understand what you mean if the script pull unfixed version.
Pls elaborate

**Reply by tteck** [reply]
The installation is set to only install version 2.1.1. Newer versions break the install.
There was a _breaking change_ by changing the installation path from `/dashy` to `/opt/dashy`.
Now please explain your issue.

**Comment by yuriw** [comment]
Ok thx
I expect when I run ‘update’ script it would pull from GitHub the fixed version 
I assumed it’s fixed and points to the new dir
What am I missing?

**Reply by yuriw** [reply]
Unless you say when need to wait for an upstream fix

**Reply by yuriw** [reply]
What I truly don't get is - why if I installed a new Dashy LXC using the latest scripts, the update still fails?!

**Reply by tteck** [reply]
I don't know how I can be clearer, **_There is currently no update path available._** 
The script **will not** update Dashy past version 2.1.1 because it breaks the install.

https://github.com/tteck/Proxmox/blob/6037d381ef3f061357b6d689f5fa89df26310c59/ct/dashy.sh#L57-L59

**Reply by yuriw** [reply]
I got that part. Thx
So how and will it be fixed at some point?

**Reply by yuriw** [reply]
@tteck how do you recommend to get it updated?

**Reply by tteck** [reply]
Wait, or switch it to Docker

**Reply by yuriw** [reply]
waiting...

**Reply by tteck** [reply]
Update _should_ be working now

**Reply by yuriw** [reply]
The update is still not working :(

**Reply by tteck** [reply]
error?

**Reply by tteck** [reply]
Works here

```
    ____             __         
   / __ \____ ______/ /_  __  __
  / / / / __  / ___/ __ \/ / / /
 / /_/ / /_/ (__  ) / / / /_/ / 
/_____/\__,_/____/_/ /_/\__, /  
                       /____/   
cat: /opt/Dashy_version.txt: No such file or directory
 ✓ Stopped Dashy
 ✓ Backed up conf.yml
 ✓ Updated Dashy to 3.0.0
 ✓ Restored conf.yml
 ✓ Cleaned
 ✓ Started Dashy
 ✓ Updated Successfully
root@dashy:~# 
```
```
    ____             __         
   / __ \____ ______/ /_  __  __
  / / / / __  / ___/ __ \/ / / /
 / /_/ / /_/ (__  ) / / / /_/ / 
/_____/\__,_/____/_/ /_/\__, /  
                       /____/   
 ✓ No update required. Dashy is already at 3.0.0
root@dashy:~# 
```

**Reply by yuriw** [reply]
![image](https://github.com/tteck/Proxmox/assets/3876474/90812114-4e8f-43a7-9327-91891b798d5b)

**Reply by tteck** [reply]
That's an issue on your end. What happened to the directory `/opt/dashy/public`?
https://github.com/tteck/Proxmox/blob/c042b7c5278f503712c628622d6da2d9c5b0301a/ct/dashy.sh#L57

**Reply by yuriw** [reply]
I don't have `/opt/dashy/public`, I have `/dashy`

I don't remember messing with them :(

What do I need to do now?

**Reply by tteck** [reply]
just `/dashy` not `/opt/dashy`?

**Reply by yuriw** [reply]
yup, nothing is in `/opt`

root@dashy:/dashy# cd /opt/
root@dashy:/opt# ls
root@dashy:/opt# 


root@dashy:/opt# cd /dashy/
root@dashy:/dashy# ls
CNAME       app.json            netlify.toml  services
Dockerfile  dist                node_modules  src
LICENSE     docker              package.json  vue.config.js
Procfile    docker-compose.yml  public        yarn.lock
README.md   docs

**Reply by tteck** [reply]
Create a new Dashy LXC

**Reply by yuriw** [reply]
I understand this way.  I was more interested in how to fix the old instance. 
:(

**Reply by tteck** [reply]
You can try this snippet to correct the path, make sure you have a backup first just in case.

```
systemctl -q --now disable dashy
mv /dashy /opt/dashy
cat <<EOF >/etc/systemd/system/dashy.service
[Unit]
Description=dashy

[Service]
Type=simple
WorkingDirectory=/opt/dashy
ExecStart=/usr/bin/yarn start
[Install]
WantedBy=multi-user.target
EOF
systemctl -q --now enable dashy
```
Afterwards, try to update.

**Reply by yuriw** [reply]
That did work and `update` after as well! :)

Thx a million @tteck !

**Comment by lissy93** [comment]
I've pushed an update (3.0.0) to Dashy which should resolve all of this.
But it does have 2 [breaking changes](https://github.com/Lissy93/dashy/discussions/1529) (change in port to `8080`, and change in config file location to `user-data/conf.yml`)

**Reply by yuriw** [reply]
I wonder if `update` can take those changes into account and fix them

**Comment by CrazyWolf13** [comment]
Dashy has just been updated to 3.0.1 which resolves a lot of auth related stuff.
Thank you for updating @tteck !

**Reply by yuriw** [reply]
@CrazyWolf13 worked like a charm!

**Comment by yuriw** [comment]
It is interesting now.
Dashy shows `Update Available: 3.1.0` but the update says ` ✓ No update required. Dashy is already at 3.0.1`

Any clues ?

**Reply by tteck** [reply]
`curl -sL https://api.github.com/repos/Lissy93/dashy/releases/latest | grep '"tag_name":' | cut -d'"' -f4` shows `3.0.1` as the latest. So does their [GitHub](https://github.com/Lissy93/dashy/releases/tag/3.0.1)

**Reply by CrazyWolf13** [reply]
I'll take care of this.

**Reply by yuriw** [reply]
@CrazyWolf13 pls drop a line here when done, TIA

**Reply by yuriw** [reply]
Any updates?

**Reply by CrazyWolf13** [reply]
Still waiting for feedback from lissy, she maintains the workflows that manage the releases.

**Reply by yuriw** [reply]
@CrazyWolf13 can you pls ping @Lissy93 ?
dashy is already on version 3.1.1

**Reply by CrazyWolf13** [reply]
Hi @yuriw 

As you already pinged her, I don't think I need to ping her again 😉
Lissy is currently really busy as the her Project was chosen for Github accelerator. 🎉

Workflows have been fixed by her, but the releases are still drafts.
As the official Docker Container and install from source is working just fine, we sadly have to give this a lower priority, as this script is not official by lissy93.

Will keep you updated.

**Reply by CrazyWolf13** [reply]
@yuriw  release is published, all should work now.

**Reply by yuriw** [reply]
Well, it does start @CrazyWolf13 and then failed:

 ✓ Stopped Dashy
 ✓ Backed up conf.yml
 / Updating Dashy to 3.1.1   
[ERROR] in line 80: exit code 0: while executing command yarn &> /dev/null

and after that dashy becomes unusable

**Reply by yuriw** [reply]
@CrazyWolf13 @tteck any suggestion on how to workaround this ^ issue?

**Reply by tteck** [reply]
The script switched from `yarn` to `npm` https://github.com/tteck/Proxmox/commit/2e268c2f49e7d40399973db07b692c2ef72392c7 https://github.com/tteck/Proxmox/commit/37f42468ba295b42f4cd636fb31d8eacbd2703ab
Create a new Dashy LXC

**Reply by yuriw** [reply]
@tteck this is not such an easy suggestion.  I will need to recreate all settings.  IMO upgrade scripts change often and every time to recreate an LXC is not a good solution.  It sacrifices the whole idea of using the tteck scripts.
Do we have steps on how to preserve all settings?

**Reply by CrazyWolf13** [reply]
> The script switched from `yarn` to `npm` [2e268c2]

@tteck 
 May I ask why this was changed?

**Reply by CrazyWolf13** [reply]
@yuriw  Why not just copy the conf.yml, it contains all your settings 😂

**Reply by tteck** [reply]
@CrazyWolf13 the script would not install Dashy using `yarn` with the new version, conflicting dependencies.

**Reply by CrazyWolf13** [reply]
@tteck I think you were supposed to add yarn --ignore-engines during install.

**Reply by yuriw** [reply]
@CrazyWolf13 
I guess you are right.  But I as an end user I rely 100% on the update script and if it has no concept of backward compatibility, then why do I need to even start using it?  I don't even know where the conf.yml is :(

I can install an LXC and install whatever I want on top of it.
I understand there could be exceptions.  But then we need exact steps on how to work around it and not just a failure error message.
If we provide an update scrip, it has to have the concept of backward compatibility.

Don't get me wrong, I love the tteck scripts and wish they were better.

**Reply by CrazyWolf13** [reply]
@yuriw  
To be honest, that's kind of the curse and the blessing, when creating such scripts there are the people that use them, to simplify and save time, and the other part is users that don't have any idea about how it works.

While the second thing is not a problem to me, you should atleast know to never inject ANY remote script from the www UNLESS you've read through it understood it and can verify it's safe to run. There's no problem, if you need to ask ChatGPT for example, no one can know everything.

If you've read through the installation script of this dashy installer, you'd see, that the config is at the following location:
`/opt/dashy/user-data/conf.yml` or at `/opt/dashy/public/conf.ym` -> https://github.com/tteck/Proxmox/blob/main/ct/dashy.sh#L68

I can understand the frustration of you, but if you feel more comfortable with a manual installation and manually updating, then go for it, but don't expect things to run smoother.- https://dashy.to/docs/deployment here some instructions on how to deploy dashy.

And I don't agree with the term end user, YOU are the syadmin, YOU are the maintainer of that dashy instance at your home, but use the awesome helper scripts to **simplify** your life.

The helper scripts are awesome, they are at the best they can be, but sometimes there are changes in the applications like dashy, where tteck needs to make quick decisions on how to continue the script, to ensure it's possible to upgrade, but sometimes that process isn't always buttery smooth, so there are times where a recreation is neccessary.



If I were you, I'd go into the old broken dashy lxc, and look for the config file, at either `/opt/dashy/user-data/conf.yml` or 
 `/opt/dashy/public/conf.ym` then cat the file with nano, paste it into any txt file on your desktop, possibly validate the copied config file here, if there were any errors that occured while copy+pasting the conf file: https://www.yamllint.com/
 
 Then create a new lxc, and paste in the config again.
 
 Hope this helps you and may clear up some questions.

**Reply by yuriw** [reply]
@CrazyWolf13 I hear you and agree - the scripts are awesome!

Does it mean they can lack backward compatibility?  Don't we want them to be better?

What does the error message `[ERROR] in line 80: exit code 0: while executing command yarn &> /dev/null` tell users?!

If update steps changed and there is no easy way to fix it, is it too much to expect to show a message and some direction on how a user can fix it instead of asking ChatGPT?

**Reply by CrazyWolf13** [reply]
@yuriw 
They shouldn't lack that, but there were breaking changes due to switch from yarn to npm, but if you complain, then why not open a feature request and submit a Pull Request yourself, get active if you want something to be changed, in the open source world a lot of people invest hours of their free-time often free of charge to provide software scripts or similar. Why not support / help them a bit.

As switching from yarn to npm are big changes I guess switching over to a new lxc makes sense.

And instead of writing an ever longer text here, why not just copy out the conf.yml, backup externally create a new lxc insert the config, it will probably take less than 5 minutes.

**Reply by yuriw** [reply]
@CrazyWolf13 I must admit that the process or installing dashy via the script and restoring the settings is easy.

I still feel that giving an informational message in a situation like this, would simply users' lives and reduce questions.

Thx for your help and good discussion :)

**Reply by CrazyWolf13** [reply]
@yuriw  
Glad to see, that you got it to work. :)
No problem.
